File: openfaas-test/go-call-c/caller/caller.go
package main

import (
  "io"
  "os"
  "fmt"
  "strings"
  "encoding/json"
  "bytes"
  "encoding/base64"
  "net/http"
  "io/ioutil"
)

type Response struct {
    Err string `json:"err"`
    Msg string `json:"msg"`
}

type Request struct {
    Msg string `json:"msg"`
}

func get_arg_from_caller() string {
  data, err := io.ReadAll(os.Stdin)
  if err != nil {
    os.Exit(0)
  }
  str := strings.TrimSpace(string(data))
  return str
}

func send_value_to_caller(output string) {
  response := Response{
    Err: "",
    Msg: output,
  }
  jsonData, _ := json.Marshal(response)

  io.WriteString(os.Stdout, string(jsonData))
}

func make_rpc(func_name string, data string) string {
  url := "http://localhost:9999/api/v1/namespaces/_/actions/"
  url += func_name
  url += "?blocking=true&result=true"
  jsonData := []byte(data)
  username := "23bc46b1-71f6-4ed5-8c54-816aa4f8c502"
  password := "123zO3xZCLrMN6v2BKK1dXYFpXlPkccOFqm12CdAsMgRU4VrNZ9lyGVCGuMDGIwP"
  auth := username + ":" + password
  encodedAuth := base64.StdEncoding.EncodeToString([]byte(auth))
  req, err := http.NewRequest("POST", url, bytes.NewBuffer(jsonData))

  req.Header.Set("Content-Type", "application/json")
  req.Header.Set("Authorization", "Basic "+encodedAuth)

  // Create a new HTTP client and send the request
  client := &http.Client{}
  resp, err := client.Do(req)
  if err != nil {
    fmt.Println("Error sending POST request:", err)
    os.Exit(0)
  }
  defer resp.Body.Close()

  // return the response body
  body, err := ioutil.ReadAll(resp.Body)
  if err != nil {
    fmt.Println("Error reading response body:", err)
    os.Exit(0)
  }
  return string(body)
}

func main() {
  input := get_arg_from_caller()
  var request Request
  err := json.Unmarshal([]byte(input), &request)
  if err != nil {
    fmt.Println("Error converting string to json:", err)
    os.Exit(0)
  }
  uuid_request := Request {
    Msg: "",
  }
  jsonData, _ := json.Marshal(uuid_request)
  resp := make_rpc("unique-id-service", string(jsonData))
  send_value_to_caller(resp)
}

File: merge_func/merge-swift-and-c/caller/caller.swift
import Foundation
import FoundationNetworking

@main
struct Function {
  struct Message: Decodable {
    let msg: String
    let err: String
  }  

  static func main() {
    let input = get_arg_from_caller()
    let json_str = ""
//    let res = make_rpc(func_name: "unique-id-service", jsonStr: json_str);
//    send_return_value_to_caller(res)
    send_return_value_to_caller("hello");
  }

  static func get_arg_from_caller() -> String {
    let inputData = FileHandle.standardInput.readDataToEndOfFile()
    guard let inputString = String(data: inputData, encoding: .utf8) else {
      return ""
    }
    return inputString.trimmingCharacters(in: .whitespacesAndNewlines)
  }

  static func send_return_value_to_caller(_ message: String) {
    print("\(message)")
  }

  static func make_rpc(func_name: String, jsonStr: String) -> String {
    var urlStr = "http://localhost:9999/api/v1/namespaces/_/actions/"
    urlStr += func_name
    urlStr += "?blocking=true&result=true"
    guard let url = URL(string: urlStr) else {
      print("Invalid URL")
      return ""
    }

    let bodyData = jsonStr.data(using: .utf8)

    // Create a POST request
    var request = URLRequest(url: url)
    request.httpMethod = "POST"
    request.setValue("application/json", forHTTPHeaderField: "Content-Type")
    request.httpBody = bodyData

    // add basic authentication
    let username = "23bc46b1-71f6-4ed5-8c54-816aa4f8c502"
    let password = "123zO3xZCLrMN6v2BKK1dXYFpXlPkccOFqm12CdAsMgRU4VrNZ9lyGVCGuMDGIwP"
    let loginString = "\(username):\(password)"
    guard let loginData = loginString.data(using: .utf8) else { return "" }
    let base64LoginString = loginData.base64EncodedString()
    request.setValue("Basic \(base64LoginString)", forHTTPHeaderField: "Authorization")

    var resp = ""
    let semaphore = DispatchSemaphore(value: 0)
    // Create a URLSession data task
    let task = URLSession.shared.dataTask(with: request) { data, response, error in
      if let res = data {
        do {
          let decoder = JSONDecoder()
          let message_err = try! decoder.decode(Message.self, from: res)
          resp = message_err.msg
        }
      }
      semaphore.signal()
    }
    

    // Start the task
    task.resume()
    semaphore.wait()
    return resp
  }
}
